flow/tasks/os/exec.go

```diff
@@ -2,129 +2,205 @@ package os
+
+	if exterr != nil {
+		return nil, fmt.Errorf("os.Exec field: value extraction failed: %w", exterr)
+	}
 
-	// todo, check failure modes, fill, not return error?
-	// (in all tasks, really)
+	// Create and configure command with goctx
+	T.cmd = exec.CommandContext(goctx, cmds[0], cmds[1:]...)
+	T.cmd.Dir = dir
+	T.cmd.Env = env
 
-	var stdout, stderr io.Writer
+	// Setup I/O
+	if stdin != nil {
+		T.cmd.Stdin = stdin
+	}
+	if stdout != nil {
+		T.cmd.Stdout = stdout
+	}
+	if stderr != nil {
+		T.cmd.Stderr = stderr
+	}
 
-	ferr := func() error {
-		ctx.CUELock.Lock()
-		defer func() {
-			ctx.CUELock.Unlock()
-		}()
-		// get and create command
-		cmds, err := extractCmd(v)
-		if err != nil {
-			return err
+	// Launch process in background
+	T.wg.Add(1)
+	go func() {
+		defer T.wg.Done()
+		defer close(processComplete)
+		defer close(processError)
+		
+		// Start the process
+		if err := T.cmd.Start(); err != nil {
+			processError <- err
+			return
 		}
-		cmd = exec.Command(cmds[0], cmds[1:]...)
 
-		// get dir / env for command
-		dir, err := extractDir(v)
+		// Wait for process completion
+		err := T.cmd.Wait()
 		if err != nil {
-			return err
+			processError <- err
 		}
-		cmd.Dir = dir
+		processComplete <- struct{}{} 
+	}()
 
-		env, err := extractEnv(v)
-		if err != nil {
-			return err
+	// Wait for completion, error, or cancellation
+	select {
+	case <-processComplete:
+		// Update return value with process state
+		if T.cmd.ProcessState != nil {
+			ret["exitcode"] = T.cmd.ProcessState.ExitCode()
+			ret["success"] = T.cmd.ProcessState.Success()
 		}
-		cmd.Env = env
+		ret, _ = fillIO(v, ret, stdout, stderr)
 
-		// setup i/o for command
-		var stdin io.Reader
-		stdin, stdout, stderr, err = extractIO(v)
-		if err != nil {
-			return err
+	case err := <-processError:
+		// Handle error according to user preferences
+		ret["error"] = err.Error()
+		if doExit {
+			return ret, err
 		}
 
-		if stdin != nil {
-			cmd.Stdin = stdin
-		}
-		if stdout != nil {
-			cmd.Stdout = stdout
-		}
-		if stderr != nil {
-			cmd.Stderr = stderr
-		}
+	case <-goctx.Done():
+		// Flow context cancelled (CTRL-C, timeout, etc)
+		T.cleanup()
+		ret["error"] = "flow cancelled"
 
-		return nil
-	}()
-	if ferr != nil {
-		return nil, ferr
-	}
+		// Capture any output that occurred before cancellation
+		ret, _ = fillIO(v, ret, stdout, stderr) 
 
-	// (TODO): check for user's abort mode preference
-	doExit, err := extractExit(v)
-	if err != nil {
-		return nil, err
+		return ret, goctx.Err()
 	}
 
-	//
-	// run command
-	//
-	rerr := cmd.Run()
-
-	// TODO, how to run in the background and wait for signal?
-
-	// build return value
-	ret := make(map[string]interface{})
+	return ret, nil
+}
 
-	//
-	// possibly fill stdout/stderr
-	//
-	ret, err = fillIO(v, ret, stdout, stderr)
-	if err != nil {
-		return nil, err
+func (T *Exec) cleanup() {
+	// Cancel our process context
+	if T.cancel != nil {
+		T.cancel()
 	}
 
-	// fill exit code / successful
```

Comment:
this was causing a segmentation fault for me. i'm kinda passed it so i forgot the reproduction. i was moving fast so i just re-wrote it with some ai help in all honesty. i've read a decent amount of the codebase and had to hand stitch this together since claude didn't exactly do well. it mainly just helped with what i would imagine a cookie cutter implementation for a nice threaded approach might look like with room for tracking multiple channels. i'm no goru but it works so far :)

Reply:
i can put it on my list to reproduce, but it might be a couple hundred thousand minutes.